File: packages/python/src/sarj_python_lint/rules/no_vague_suppression_description.py
from __future__ import annotations

from pathlib import PurePosixPath
import re
from typing import TYPE_CHECKING, ClassVar, final, override

from sarj_python_lint.rule_base import (
    AutofixPolicy,
    ColumnEncoding,
    Diagnostic,
    ExampleFile,
    ExampleOutcome,
    Rule,
    RuleCategory,
    RuleDocumentation,
    RuleExample,
    Severity,
)
from sarj_python_lint.rules._paths import is_generated
from sarj_python_lint.rules._suppression_comments import scan_comments_or_none


if TYPE_CHECKING:
    from pathlib import Path

    from sarj_python_lint.rules._suppression_comments import Comment


_INLINE_DIRECTIVE = (
    r"(?:noqa\s*:\s*[A-Z0-9_, -]+|sarj-noqa\s*:\s*[A-Z0-9_, -]+|"
    r"type\s*:\s*ignore\s*\[[^\]]+\]|(?:based)?pyright\s*:\s*ignore\s*\[[^\]]+\]|"
    r"ruff\s*:\s*ignore\s*\[[^\]]+\])"
)
_FILE_DIRECTIVE = (
    r"(?:ruff\s*:\s*(?:file-ignore\s*\[[^\]]+\]|noqa(?:\s*:\s*[A-Z0-9_, -]+)?)|"
    r"flake8\s*:\s*noqa(?:\s*:\s*[A-Z0-9_, -]+)?)"
)
_DIRECTIVE_START = rf"(?:{_INLINE_DIRECTIVE}|{_FILE_DIRECTIVE})"
_DIRECTIVE_RE = re.compile(
    rf"(?:^|#\s*)(?:(?P<inline>{_INLINE_DIRECTIVE})|(?P<file>{_FILE_DIRECTIVE}))"
    rf"\s*(?:—|–|--)\s*(?P<description>.+?)(?=\s+#\s*{_DIRECTIVE_START}|\s*$)",
    re.IGNORECASE,
)
_VAGUE_RE = re.compile(
    r"^(?:(?:(?:because|just)\s+)?(?:needed|required|necessary|intentional(?:ly)?|ignore(?:d)?|"
    r"false[- ]positive|type error|typing issue|lint issue|known issue|expected|safe|harmless|todo|"
    r"upstream bug|tool limitation)(?:\s+(?:here|for now|workaround))?|"
    r"(?:temporary\s+)?workaround|by design|not an issue|fix later|pre-existing,? out of scope|"
    r"(?:to satisfy|required by)\s+(?:the\s+)?(?:tool|linter|type checker|mypy|pyright|ruff)|"
    r"python|mypy|pyright|ruff)[.!?]?$",
    re.IGNORECASE,
)


@final
class NoVagueSuppressionDescription(Rule):
    id = "no-vague-suppression-description"
    code = "SARJ419"
    documentation: ClassVar[RuleDocumentation | None] = RuleDocumentation(
        summary="Generic suppression descriptions do not make the exception auditable.",
        rationale="Generic reasons satisfy review conventions without making the suppressed risk auditable or removable.",
        remediation=(
            "Name the concrete tool mismatch or safety invariant and the condition that would allow removal. Cite an "
            "upstream issue or affected version when that is the durable boundary."
        ),
        category=RuleCategory.MAINTAINABILITY,
        autofix=AutofixPolicy.NONE,
        limitations=(
            "Only supported scoped Python suppression directives whose description exactly matches a bounded generic phrase are checked.",
            "Missing descriptions, blanket directives, formatter and coverage pragmas, Pylint directives, and whether a suppression is necessary remain outside this rule.",
            "Ordinary scoped directives must be attached to code; recognized Ruff and Flake8 file directives may stand alone.",
        ),
        examples=(
            RuleExample(
                example_id="generic-suppression-reason",
                title="Generic reason does not make the suppression auditable",
                outcome=ExampleOutcome.MATCH,
                files=(
                    ExampleFile.python(
                        "adapter.py", "value = vendor.value  # ruff: ignore[attr-defined] -- false positive\n"
                    ),
                ),
                focus_path=PurePosixPath("adapter.py"),
                expected_count=1,
                public=True,
            ),
            RuleExample(
                example_id="concrete-suppression-reason",
                title="Concrete runtime mismatch explains the suppression",
                outcome=ExampleOutcome.NO_MATCH,
                files=(
                    ExampleFile.python(
                        "adapter.py",
                        "value = vendor.value  # type: ignore[attr-defined] -- vendor stubs omit the runtime field\n",
                    ),
                ),
                focus_path=PurePosixPath("adapter.py"),
                expected_count=0,
                public=True,
            ),
        ),
    )
    description = documentation.summary

    @override
    def check(self, path: Path, source: str) -> list[Diagnostic]:
        if is_generated(path, source):
            return []
        comments = scan_comments_or_none(source)
        if comments is None:
            return []
        diagnostics: list[Diagnostic] = []
        for comment in comments:
            vague_reason = _vague_reason(comment)
            if vague_reason is None:
                continue
            diagnostics.append(
                Diagnostic(
                    path=path,
                    line=comment.line,
                    col=comment.col,
                    code=self.code,
                    message=(
                        f'Suppression reason "{vague_reason}" is generic; name the concrete mismatch or safety invariant.'
                    ),
                    severity=Severity.WARNING,
                    column_encoding=ColumnEncoding.CODEPOINTS,
                )
            )
        return diagnostics


def _vague_reason(comment: Comment) -> str | None:
    for match in _DIRECTIVE_RE.finditer(comment.body):
        reason = match["description"].strip()
        if (not comment.standalone or match["file"] is not None) and _VAGUE_RE.fullmatch(reason) is not None:
            return reason
    return None

File: packages/typescript/src/rules/no-vague-suppression-description.ts
/**
 * @fileoverview no-vague-suppression-description — a suppression reason must name the concrete mismatch or invariant, not merely claim the suppression is needed.
 *
 * Examples: https://github.com/sarj-ai/code-standards/blob/main/packages/typescript/tests/rules/no-vague-suppression-description.test.ts
 */

import { createRule, type RuleDocumentation } from "./_docs.js";
import { isGeneratedFile } from "./_paths.js";

type MessageIds = "vagueDescription";
type Options = readonly [];

const DIRECTIVE_WITH_DESCRIPTION_RE = /^(?:eslint-(?:disable|disable-next-line|disable-line)\b[^:\n]*?|@ts-expect-error\b)\s*(?::|--)\s*(.+?)\s*$/iu;
const VAGUE_RE = /^(?:needed|required|intentional(?:ly)?|ignore(?:d)?|false positive|type error|typescript|to satisfy (?:the )?(?:linter|typescript|type checker))\.?$/iu;

export const noVagueSuppressionDescriptionDocumentation = {
  summary:
    "Require suppression descriptions to name the concrete mismatch or invariant instead of a generic non-reason.",
  rationale:
    "Generic phrases satisfy require-description mechanically while leaving reviewers unable to audit the risk or remove stale debt.",
  remediation:
    "Name the exact type/runtime mismatch, external contract, or safety invariant that makes this suppression acceptable.",
  category: "maintainability",
  limitations: [
    "Only ESLint disable comments and TypeScript expect-error directives are checked.",
    "The rule uses a small anchored vocabulary and does not score prose quality generally.",
    "Generated files and descriptions containing any concrete context are excluded.",
  ],
  examples: [
    {
      id: "concrete-runtime-mismatch",
      title: "Explain the concrete runtime contract",
      outcome: "no-match",
      files: [
        {
          path: "src/adapter.ts",
          source:
            "// @ts-expect-error -- vendor types omit the runtime requestId field\nreturn response.requestId;",
        },
      ],
      focusPath: "src/adapter.ts",
      expectedCount: 0,
      public: true,
    },
    {
      id: "generic-suppression-reason",
      title: "Reject a suppression with no auditable reason",
      outcome: "match",
      files: [
        {
          path: "src/adapter.ts",
          source: "// @ts-expect-error -- false positive\nreturn response.requestId;",
        },
      ],
      focusPath: "src/adapter.ts",
      expectedCount: 1,
      public: true,
    },
  ],
} as const satisfies RuleDocumentation;

export default createRule<Options, MessageIds>({
  name: "no-vague-suppression-description",
  documentation: noVagueSuppressionDescriptionDocumentation,
  meta: {
    type: "suggestion",
    docs: {
      description:
        "Require suppression descriptions to name the concrete mismatch or invariant instead of a generic non-reason.",
    },
    schema: [],
    messages: {
      vagueDescription:
        "Suppression description `{{description}}` does not explain why the suppressed diagnostic is safe here. Name the concrete type/runtime mismatch, external contract, or invariant.",
    },
  },
  defaultOptions: [],
  create(context) {
    if (isGeneratedFile(context.filename, context.sourceCode.text)) {
      return {};
    }
    return {
      Program(): void {
        for (const comment of context.sourceCode.getAllComments()) {
          const text = comment.value.trim();
          const description = DIRECTIVE_WITH_DESCRIPTION_RE.exec(text)?.[1]?.trim();
          if (description === undefined || !VAGUE_RE.test(description)) continue;
          context.report({
            loc: comment.loc,
            messageId: "vagueDescription",
            data: { description },
          });
        }
      },
    };
  },
});
